File: src/main/java/com/telegram/folobot/service/FoloPidorService.java
package com.telegram.folobot.service;

import com.telegram.folobot.domain.FoloPidorEntity;
import com.telegram.folobot.domain.FoloPidorId;
import com.telegram.folobot.dto.FoloPidorDto;
import com.telegram.folobot.mappers.FoloPidorMapper;
import com.telegram.folobot.repos.FoloPidorRepo;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;
import java.util.SplittableRandom;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class FoloPidorService {
    private final FoloPidorRepo foloPidorRepo;
    private final FoloPidorMapper foloPidorMapper;

    /**
     * Прочитать все
     * @return {@link List<FoloPidorDto>}
     */
    public List<FoloPidorDto> findAll() {
        return foloPidorRepo.findAll(
                Sort.by("id.chatId").ascending()
                        .and(Sort.by("score").descending())).stream()
                .map(foloPidorMapper::mapToFoloPidorDto)
                .toList();
    }

    /**
     * Чтение по ключу
     * @param chatId Id чата
     * @param userId Id пользователя
     * @return {@link FoloPidorDto}
     */
    public FoloPidorDto findById(Long chatId, Long userId) {
        return foloPidorRepo.findById(new FoloPidorId(chatId, userId))
                .map(foloPidorMapper::mapToFoloPidorDto)
                .orElse(new FoloPidorDto(chatId, userId).setNew(true));
    }

    /**
     * Проверка существования по ключу
     * @param chatId Id чата
     * @param userId Id пользователя
     * @return да/нет
     */
    public boolean existsById(Long chatId, Long userId) {
        return foloPidorRepo.existsById(new FoloPidorId(chatId, userId));
    }

    /**
     * Получение по Id чата
     * @param chatId Id чата
     * @return {@link List<FoloPidorDto>}
     */
    public List<FoloPidorDto> findByIdChatId(Long chatId) {
        return foloPidorRepo.findByIdChatId(chatId,
                        Sort.by("id.chatId").ascending()
                                .and(Sort.by("score").descending())).stream()
                .map(foloPidorMapper::mapToFoloPidorDto)
                .toList();
    }

    /**
     * Выбор случайного фолопидора
     *
     * @param chatid ID чата
     * @return {@link FoloPidorEntity}
     */
    public FoloPidorDto getRandom(Long chatid) {
        //Получаем список фолопидоров для чата
        List<FoloPidorEntity> foloPidorEntities = foloPidorRepo.findByIdChatId(chatid);
        //Выбираем случайного
        return foloPidorMapper.mapToFoloPidorDto(foloPidorEntities
                .get(new SplittableRandom().nextInt(foloPidorEntities.size())));
    }

    /**
     * Получение топ 10 фолопидоров чата
     * @param chatId Id чата
     * @return {@link List<FoloPidorDto>}
     */
    public List<FoloPidorDto> getTop(Long chatId) {
        return foloPidorRepo.findByIdChatId(chatId)
                .stream()
                .map(foloPidorMapper::mapToFoloPidorDto)
                .filter(FoloPidorDto::hasScore)
                .collect(
                        Collectors.groupingBy(Function.identity(),
                                Collectors.summingInt(FoloPidorDto::getScore)))
                .entrySet().stream()
                .limit(10)
                .map(e -> e.getKey().setScore(e.getValue()))
                .sorted(Comparator.comparing(FoloPidorDto::getScore).reversed())
                .toList();
    }

    /**
     * Сохранение
     * @param dto {@link FoloPidorDto}
     */
    public void save(FoloPidorDto dto) {
        foloPidorRepo.save(foloPidorMapper.mapToFoloPidorEntity(dto));
    }

    /**
     * Удаление
     * @param dto {@link FoloPidorDto}
     */
    public void delete(FoloPidorDto dto) {foloPidorRepo.delete(foloPidorMapper.mapToFoloPidorEntity(dto)); }
}

File: src/main/kotlin/com/telegram/folobot/service/FoloPidorService.kt
package com.telegram.folobot.service

import com.telegram.folobot.model.dto.FoloPidorDto
import com.telegram.folobot.model.dto.toEntity
import com.telegram.folobot.persistence.entity.FoloPidorId
import com.telegram.folobot.persistence.entity.toDto
import com.telegram.folobot.persistence.repos.FoloPidorRepo
import org.springframework.stereotype.Component
import java.time.LocalDate

@Component
class FoloPidorService(
    private val foloPidorRepo: FoloPidorRepo,
    private val userService: UserService
) {

    /**
     * Прочитать все
     * @return [<]
     */
    fun findAll(): List<FoloPidorDto> {
        return foloPidorRepo.findAll()
            .map { it.toDto() }
            .sortedWith(compareBy<FoloPidorDto> { it.id.chatId }.thenByDescending { it.score })
    }

    /**
     * Чтение по ключу
     * @param chatId Id чата
     * @param userId Id пользователя
     * @return [FoloPidorDto]
     */
    fun findById(chatId: Long, userId: Long): FoloPidorDto {
        return foloPidorRepo.findFoloPidorById(FoloPidorId(chatId, userId))?.toDto()
            ?: FoloPidorDto(chatId, userId)
    }

    /**
     * Проверка существования по ключу
     * @param chatId Id чата
     * @param userId Id пользователя
     * @return да/нет
     */
    fun existsById(chatId: Long, userId: Long): Boolean {
        return foloPidorRepo.existsById(FoloPidorId(chatId, userId))
    }

    /**
     * Получение по Id чата
     * @param chatId Id чата
     * @return [<]
     */
    fun findByIdChatId(chatId: Long): List<FoloPidorDto> {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .sortedWith(compareBy<FoloPidorDto> { it.id.chatId }.thenByDescending { it.score })
    }

    /**
     * Выбор случайного фолопидора
     *
     * @param chatId ID чата
     * @return [FoloPidorDto]
     */
    fun getRandom(chatId: Long): FoloPidorDto {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .filter { it.isAnchored() || (it.isValid() && userService.isInChat(it, chatId)) }
            .random()
    }

    /**
     * Получение топ 10 фолопидоров чата
     * @param chatId Id чата
     * @return [<]
     */
    fun getTop(chatId: Long): List<FoloPidorDto> {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .filter { it.isValidTop() }
            .sortedWith(compareByDescending<FoloPidorDto> { it.score }.thenByDescending { it.lastWinDate })
            .take(10)
    }

    /**
     * Паассивные фолопидоры чата
     * @param chatId Id чата
     * @return [<]
     */
    fun getSlackers(chatId: Long): List<FoloPidorDto> {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .filter { userService.isInChat(it, chatId) && it.isValidSlacker() }
            .sortedBy { it.lastActiveDate }
            .take(10)
    }

    /**
     * Получение списка андердогов
     * @param chatId Id чата
     * @return list of [FoloPidorDto]
     */
    fun getUnderdogs(chatId: Long): List<FoloPidorDto> {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .filter { userService.isInChat(it, chatId) && it.isValidUnderdog() }
    }

    /**
     * Активные фолопидоры чата
     * @param chatId Id чата
     * @return [<]
     */
    fun getTopActive(chatId: Long): List<FoloPidorDto> {
        return foloPidorRepo.findByIdChatId(chatId)
            .map { it.toDto() }
            .filter { it.lastActiveDate == LocalDate.now() }
            .sortedByDescending { it.messagesPerDay }
            .take(5)
    }

    /**
     * Сохранение
     * @param dto [FoloPidorDto]
     */
    fun save(dto: FoloPidorDto) {
        foloPidorRepo.save(dto.toEntity())
    }

    /**
     * Удаление
     * @param dto [FoloPidorDto]
     */
    fun delete(dto: FoloPidorDto) {
        foloPidorRepo.delete(dto.toEntity())
    }
}